src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp

```diff
@@ -24,18 +26,87 @@
 #include "Parallel/Reduction.hpp"
 #include "ParallelAlgorithms/Interpolation/InterpolationTargetDetail.hpp"
 #include "ParallelAlgorithms/Interpolation/Protocols/PostInterpolationCallback.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
 #include "Utilities/Functional.hpp"
+#include "Utilities/GetOutput.hpp"
 #include "Utilities/Gsl.hpp"
+#include "Utilities/MakeString.hpp"
 #include "Utilities/PrettyType.hpp"
 #include "Utilities/ProtocolHelpers.hpp"
 #include "Utilities/TMPL.hpp"
 
 namespace intrp {
 namespace callbacks {
+namespace detail {
+// Fills the legend and row of spherical harmonic data to write to disk
+//
+// The number of coefficients to write is based on `max_l`, the maximum value
+// that the input `strahlkorper` could possibly have. When
+// `strahlkorper.l_max() < max_l`, coefficients with \f$l\f$ higher than
+// `strahlkorper.l_max()` will simply be zero. Assuming the same `max_l` is
+// always used for a given surface, we will always write the same number of
+// columns for each row, as `max_l` sets the number of columns to write
+template <typename Frame>
+void fill_ylm_legend_and_data(
+    const gsl::not_null<std::vector<std::string>*> legend,
+    const gsl::not_null<std::vector<double>*> data,
+    const ylm::Strahlkorper<Frame>& strahlkorper, const double time,
+    const size_t max_l) {
+  ASSERT(max_l >= strahlkorper.l_max(),
+         "The Lmax of the Ylm data to write ("
+             << strahlkorper.l_max()
+             << ") is larger than the maximum value that l can be (" << max_l
+             << ").");
+  const std::array<double, 3> expansion_center =
+      strahlkorper.expansion_center();
+  const std::string frame{get_output(Frame{})};
+  // we only store half and thus only write half of the coefficients
+  const size_t num_coefficients =
+      ylm::Spherepack::spectral_size(max_l, max_l) / 2;
```

Comment:
I'm still not quite sure this number is correct (I find the Strahlkorper/Spherepack docs a bit confusing). You do a `resize()` of the data below where you fill it with zeros, so it needs to be correct. Please add a test where `max_l` is larger than `strahlkorper.l_max()`.

Reply:
Ah so if you look at the documentation for `Strahlkorper::coefficients()`, my understanding is that we store half of the Spherepack coefficients (i.e. half of `spectral_size()`), so that is what is written. We could also dump everything (the whole, not just the half) but then we would just write a bunch of zeros (the other half, whose actual values it seems we don't store due to the simple complex conjugation relationship described in the docs I mentioned). `Test_ReadSurfaceData.cpp` tests this filling function for `max_l > strahlkorper.l_max()` but let me know if you want me to add another test or add additional documentation!

Reply:
And please let me know if you have a different understanding of those docs!

Reply:
Apologies - I think it's not actually half, but I think what I originally had in my original commit is correct: (max_l + 1)^2. I am not able to check this now but I will, just wanted to amend my above statement. Essentially, I think the number of coefficients is the sum of the odd numbers from
1 to (max_l + 1), which is (max_l + 1)^2. I will check this again, apologies!

Reply:
The result of `ylm::Spherepack::spectral_size() / 2` is `(l_max + 1)^2` and I agree with your statement. So I think all is good. Just add a test where `l_max` is larger than `strahlkorper.l_max()` to `Test_ObserveTimeSeries` (in addition to the one in `Test_ReadSurfaceData`).